### DISCUSSION TITLE
How do I set the Binding property in <Application.Styles> through DataContext?

### DISCUSSION THREAD

**Discussion by xinghaiyege** [AvaloniaUI/Avalonia#10542] [Q&A] [answered]
Avalonia version: 11 preview-5

I want to set the theme color of the app by loading the configuration.
But setting DataContext in App.axaml.cs cannot set the Binding data of <Application.Styles>.

**Comment by timunie** [comment]
Can you show more code?

**Comment by xinghaiyege** [comment]
App.axaml
```xaml
<Application xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             x:Class="Avalonia_Test.App" RequestedThemeVariant="Light">
	<Application.Styles>
		<FluentTheme/>
		<Style Selector="TextBlock">
			<Setter Property="Foreground" Value="{Binding Color1}"/>
			<Setter Property="Background" Value="{Binding Color2}"/>
			<Setter Property="Padding" Value="20,16"/>
		</Style>
	</Application.Styles>
</Application>
```

**Reply by timunie** [reply]
I assume the TextBlock has a DataContext which has Color1 and Color2 available? In that case just add `x:DataType` to the `Style`-tag. If not, specify the `Binding Source`, for Example: 

```xml
<Setter Property="Foreground" Value="{Binding Source={x:Static MyClass.Instance}, Path=Color1}"/>
```

**Comment by xinghaiyege** [comment, marked as answer]
I tried it, it seems that it can only have an effect on `Window`, not on `Application`.

I missed a key piece of information, the `Avalonia 11 Preview-5` I used.

**Reply by timunie** [reply]
That's just not true. The Binding is pointing to the DataContext defined. App can have a DataContext, if you have defined it. But anyway, the Style will have the DataContext of the TextBlock which got the Binding applied. 

Note: 1. Color1 and Color2 needs to be a Brush or use a Converter. 

Another idea would be to use DynamicResource, which is common if the colors are the same for all TextBlocks

**Reply by xinghaiyege** [reply]
Your other idea is correct. After testing, only `DynamicResource` and `StaticResource` work on `Application`.

**Comment by EmVee381** [comment]
I was basically solving the same thing recently and stumbled upon your discussion, but I got fixated on wanting the binding to be directly to the MainViewModel. This is because I'm using Xplat and loading Settings via a service that DI passes into the ViewModel. I have a different service for each system, that's why I do it this way, and I load the Android font size scale factor.

Maybe I didn't explain it perfectly... I do it this way because I have a router in MainView for nested screens, and I wanted the styles to apply to all nested views. It's like application-level styling, but starting from MainView and going down the visual tree.

Eventually, I came up with another solution that I want to share here. Maybe I'm doing it wrong, but it works and I'm satisfied.

On the UserControl I set x:Name and in the setter I bind:
```xml
<Setter Property="FontSize"
        Value="{Binding #MainViewObject.DataContext.DefaultFontSize}" />
```

I know it might seem a bit weird to reference back to the MainView from within itself using the #MainViewObject name, but this bypasses the binding issues from nested views so they can reach the correct property.

Maybe it will help someone :)

**Reply by stevemonaco** [reply]
For these cases, I strongly prefer using inherited attached properties. Attach the property to your root view and it will be pushed down through the entire visual tree (ie. available for binding). This is how `TextElement.FontSize` already works:

https://github.com/AvaloniaUI/Avalonia/blob/ada1e72ce8dce25bfdf58fe5f848d7f6652e7b00/src/Avalonia.Controls/Documents/TextElement.cs#L34-L42

I use this approach for a lot of things -- responsive sizing, routing, etc. You have to be careful about memory leaks as you have children referencing a potentially longer-living parent, but it easily solves app-specific plumbing issues.
